packages/mdc-shape/foundation.js

```diff
@@ -0,0 +1,419 @@
+      this.penumbraShadowOffsetY_ = this.finalPenumbraShadowOffsetY_;
+      this.umbraShadowBlur_ = this.finalUmbraShadowBlur_;
+      this.umbraShadowOffsetY_ = this.finalUmbraShadowOffsetY_;
+      this.umbraShadowSpread_ = this.finalUmbraShadowSpread_;
+      finished = true;
+    }
+
+    this.redraw();
+
+    if (!finished && this.elevating_) {
+      this.animationFrameId_ = requestAnimationFrame(() => this.elevate_());
+    } else {
+      this.elevating_ = false;
+    }
+  }
+
+  clear_() {
+    this.ctx_.clearRect(0, 0, this.adapter_.getCanvasWidth(), this.adapter_.getCanvasHeight());
+  }
+
+  draw_() {
+    const devicePixelRatio = this.adapter_.getDevicePixelRatio();
+    const noSpreadPath = this.generateShadowPath_(0);
+    this.drawAmbientShadow_(devicePixelRatio, noSpreadPath);
+    this.drawPenumbraShadow_(devicePixelRatio, noSpreadPath);
+    this.drawUmbraShadow_(devicePixelRatio);
+    this.drawColoredShape_(noSpreadPath);
+    this.adapter_.setClipPathData(this.generateClipPath_());
+  }
+
+  drawAmbientShadow_(devicePixelRatio, noSpreadPath) {
+    this.ctx_.fillStyle = '#FFF';
+    this.ctx_.shadowColor = 'rgba(0, 0, 0, 0.12)';
+    this.ctx_.shadowBlur = this.ambientShadowBlur_ * devicePixelRatio;
+    this.ctx_.shadowOffsetY = this.ambientShadowOffsetY_ * devicePixelRatio;
+    this.ctx_.fill(this.adapter_.createPath2D(noSpreadPath));
+  }
+
+  drawPenumbraShadow_(devicePixelRatio, noSpreadPath) {
+    this.ctx_.shadowColor = 'rgba(0, 0, 0, 0.14)';
+    this.ctx_.shadowBlur = this.penumbraShadowBlur_ * devicePixelRatio;
+    this.ctx_.shadowOffsetY = this.penumbraShadowOffsetY_ * devicePixelRatio;
+    this.ctx_.fill(this.adapter_.createPath2D(noSpreadPath));
+  }
+
+  drawUmbraShadow_(devicePixelRatio) {
+    const umbraPath = this.generateShadowPath_(this.umbraShadowSpread_);
+    this.ctx_.shadowColor = 'rgba(0, 0, 0, 0.2)';
+    this.ctx_.shadowBlur = this.umbraShadowBlur_ * devicePixelRatio;
+    this.ctx_.shadowOffsetY = this.umbraShadowOffsetY_ * devicePixelRatio;
+    this.ctx_.fill(this.adapter_.createPath2D(umbraPath));
+  }
+
+  drawColoredShape_(noSpreadPath) {
+    this.ctx_.fillStyle = this.background_;
+    this.ctx_.shadowColor = 'rgba(0, 0, 0, 0)';
+    this.ctx_.shadowBlur = 0;
+    this.ctx_.shadowOffsetY = 0;
+    this.ctx_.fill(this.adapter_.createPath2D(noSpreadPath));
+  }
+
+  get finalAmbientShadowOffsetY_() {
+    return AMBIENT_OFFSET_Y[this.elevation_];
+  }
+
+  get finalAmbientShadowBlur_() {
+    return AMBIENT_BLUR[this.elevation_];
+  }
+
+  get finalPenumbraShadowOffsetY_() {
+    return this.elevation_;
+  }
+
+  get finalPenumbraShadowBlur_() {
+    return PENUMBRA_BLUR[this.elevation_];
+  }
+
+  get finalUmbraShadowOffsetY_() {
+    return UMBRA_OFFSET_Y[this.elevation_];
+  }
+
+  get finalUmbraShadowBlur_() {
+    return UMBRA_BLUR[this.elevation_];
+  }
+
+  get finalUmbraShadowSpread_() {
+    return UMBRA_SPREAD[this.elevation_];
+  }
+}
+
+const UMBRA_OFFSET_Y = [
```

Comment:
Why are these constants here rather than in a constants file? Also, are these more directly related to elevation rather than shapes? I suspect these correspond to styles in mdc-elevation/_variables.scss and it would be *really* easy for these to get out of sync with those, being in a completely separate package.

Is there any other stuff here that should be considered for mdc-elevation instead? Should shapes be dependent on elevation?

Reply:
Plus One

Reply:
Moved it to mdc-elevation!